### DISCUSSION TITLE
How to find the Tile Height / Width of an existing Pyramidal Image

### DISCUSSION THREAD

**Discussion by michael-scott-phoseon** [libvips/libvips#2478] [Q&A]
We are getting pre-created pyramidal OME.Tiff files from a 3rd party containing 5 pages, 3 of which are pyramidal.  These pyramidal images are too large for LibVip to load as jpeg, or bitmap (so NetVip.ToBitmap doesn't work either).  So now I need to look at loading each page via it's Tiles.  However I only see examples of CREATING tiled images, not reading them.  How does one find out the tile size of a page in a pyramidal tile structure.

We are using this sample for now, but the pages in the files we are getting from the third party are much larger (hence our problem).

https://downloads.openmicroscopy.org/images/OME-TIFF/2016-06/sub-resolutions/Fluorescence/LuCa-7color_Scan1.ome.tiff

Or... for Pyramidal OME.Tiff files, do we just store each page of the pyramid in the tiff file, but not the tiles that make it up?  If so, how ca we determine the tile height / width?  Make it up, and create the tiles themselves?

Michael

**Comment by jcupitt** [comment]
Hi @michael-scott-phoseon,

You just open the image and read out chunks of it, it should be really easy. I think you need to explain what you're trying to accomplish in more detail. Are you making an image viewer?

**Comment by michael-scott-phoseon** [comment]
John,

Yes we are.  We will load the lowest resolution image (top of the pyramid) into a WPF window, then provide a slider to move back and forth between each resolution in the pyramid. Then we will place the slide overview and slide label elsewhere on the screen

**Comment by jcupitt** [comment]
I'm actually working on an image viewer at the moment, though for linux.

https://github.com/jcupitt/vipsdisp

It's on flatpak, if you have a linux machine handy:

https://flathub.org/apps/details/org.libvips.vipsdisp

I aim to make a win build at some point, though not yet.

In your case, I would make a 500 x 500 pixel (for example) window and put a couple of scrollbars by it. When the user moves the scrollbars, fetch a new 500x500 block of pixels from the image using netvips and paint it on to the screen. Because you never fetch the whole image, it should be quick, and won't need much memory. And it's just a fixed image window and two scrollbars, so it should be simple to build.

To get fancier, you could use `vips_sink_screen()`. This is an asynchronous operation, so it'll do the fetch in the background.

**Reply by michael-scott-phoseon** [reply]
John,

Yes, I am making an Image Viewer (think Google Earth) among other things in C#.  Our OME-Tiff files contain 5 images.  A Slide Label, a Slide Overview, and a 3 layer Pyramidal Image with a Low / Medium / High Res image.  QuPath, Fiji, and Gimp do not show any Tiles (but perhaps I don't know what I am looking for), so I think it is literally 3 images.  The High Res image is Image 0, and it uses subIFD's for the other 2 images.  I can load all three via the following:

     imageHighRes  = NetVips.Image.Tiffload(SlideName); // High Res
     imageLowRes  = NetVips.Image.Tiffload(SlideName, subifd: 0); // Low Res
     imageMedRes  = NetVips.Image.Tiffload(SlideName, subifd: 1); // Med Res

But trying to extract the actual image for imageHighRes is the problem.  I need to eventually save the image in a WPF Image, BitmapSource, or ImageSource, but just trying to save it as a regular Windows Forms Bitmap fails.  I even tried using:

     Bitmap bmp = NetVips.Extensions.BitmapConverter.ToBitmap(imageHighRes);

but it leads to a call similar to 

     Bitmap tst3 = new Bitmap(width, height, PixelFormat.Format24bppRgb);

and that's just too big.  So what I'm looking for is sample code that breaks imageHighRes into tiles.  can you point me in the right direction?

**Reply by jcupitt** [reply]
The simplest thing would just be to crop out the bits you need. Something like:

```
image = NetVips.Image.Tiffload(SlideName);
tile = image.Crop(0, 0, 1000, 1000)
mem = tile.WriteToMemory()
bitmap = new Bitmap(mem, ...)
```

And now you can display the 1000 x 1000 pixel window at the top-left. When the user scrolls with your scrollbars, you need to crop and display a new area. libvips does a lot of caching behind the scenes, so this is actually quite quick (use `image` for each crop, just redo the crop and WriteToMemory). When the user zooms, pick a different base image. You can use subsample / zoom to scale the image to intermediate zoom levels.

This should be really easy (100 lines of code?), and perform reasonably well.

You can get a stage fancier and use regions instead. This will make better use of caches, though you'll lose threading.

The gold standard is fancier again: use regions, but render them in a set of background threads, and paint the screen asynchronously as your background worker threads produce computed tiles. My viewer above ^^^ does this, and also exploits the GPU for the scaling step.

**Reply by michael-scott-phoseon** [reply]
I will check this out in a few minutes, but what about image.Tilecache?  What does it do?  I'm confused between NetVips.Tilecache vs. libvips.tilecache vs. vips-tilecache.  Care to elaborate?

**Reply by jcupitt** [reply]
There's one thing (`tilecache`) which you can call from C with `vips_tilecache()` or C# with `image.Tilecache()`.

My viewer above has a class called `Tilecache`, confusingly :( perhaps I should have picked a different name :(
